### PR TITLE
[WIP] AcceptAsync with existing Socket on Unix

### LINKED ISSUE #1483: AcceptAsync with existing Socket not implemented on Unix
Windows has built-in support for accepting into an existing socket, something not available on Unix.  But we can approximate it in the .NET layer on Unix, e.g. by accepting into a new socket and transferring over the SafeCloseHandle and related state.  This will allow the same APIs to be used cross-platform so that code running on Windows gets the intended benefits while not having to fork based on the OS.

### PR DESCRIPTION
**Pull request by marcusturewicz**
Submitting initially to get feedback before continuing:
- Allowed non-null acceptSocket in Unix
- Copied the existing socket's state to the created socket after it has been accepted.

TODO
- [ ] Finalise implementation
- [ ] Tests

cc @stephentoub 

Fixes #1483

### CODE REVIEW

**Comment by marcusturewicz** [comment]
Is this looking ok, or am I completely off track?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 1]
```diff
@@ -3665,6 +3665,32 @@ public IAsyncResult BeginAccept(Socket? acceptSocket, int receiveSize, AsyncCall
...
+                asyncResult.AcceptSocket!._nonBlockingConnectInProgress = acceptSocket._nonBlockingConnectInProgress;
+                asyncResult.AcceptSocket!._nonBlockingConnectRightEndPoint = acceptSocket._nonBlockingConnectRightEndPoint;
+                asyncResult.AcceptSocket!._protocolType = acceptSocket._protocolType;
+                asyncResult.AcceptSocket!._receivingPacketInformation = acceptSocket._receivingPacketInformation;
+                asyncResult.AcceptSocket!._remoteEndPoint = acceptSocket._remoteEndPoint;
+                asyncResult.AcceptSocket!._rightEndPoint = acceptSocket._rightEndPoint;
+                asyncResult.AcceptSocket!._socketType = acceptSocket._socketType;
+                asyncResult.AcceptSocket!._willBlock = acceptSocket._willBlock;
+                asyncResult.AcceptSocket!._willBlockInternal = acceptSocket._willBlockInternal;
+            }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:3692]: A few thoughts:
- This should only be done on Unix
- The temporary socket needs to be disposed of in some way, but in a way that avoids it trying to close it's handle.  Might be best / easiest to do a GC.SuppressFinalize on it.
- Similarly the destination socket's handle probably needs to be closed prior to overwriting everything, but the semantics of that will need to be validated against the Windows behavior
- I'd need to double-check, but I believe there are multiple accept code paths, with this only covering one of them, e.g. I think there's another in SocketAsyncEventArgs.
- You might want to start by writing a bunch of tests that pass on Windows, and then getting the Unix implementation to pass them as well

**Comment by stephentoub** [comment]
@marcusturewicz, are you planning to continue working on this soon?  If not, that's fine, I'll just close it in the interim.

**Comment by marcusturewicz** [comment]
I’d like to continue it, but I’m not clear on what to do. Someone with more knowledge of sockets might be more successful.

**Comment by stephentoub** [comment]
Ok. Thanks very much for your interest and initial attempt! I'm going to go ahead and close it.

**Pull request closed without merging**